### DISCUSSION TITLE
Investigate some electron alternatives using system webviews

### DISCUSSION THREAD

**Discussion by arthuro555** [4ian/GDevelop#2546] [Enhancement ideas (and their technical discussions)]
It could be be interesting for some users to have lighter binaries, which would be possible by providing some additional export types for webview based technologies like Tauri or deno-webview, allowing the wrapper app to be in the case of Tauri sometimes as light as 600kb, way different from the ~100mb from electron.

Electron do is a more reliable and battle tested technology, so I am not saying we should make a full switch, just that providing an alternative would be nice.

**Comment by Bouh** [comment]
If we go from 100mb with Electron to 600kb with Tauri what do we lose?
Is it possible to port all the APIs we use on Electron to Tauri?

**Reply by 4ian** [reply]
The question is indeed what do we lose, what do we gain?
Taurai/alternative lightweight "webviews" have an advantage in terms of size and maybe memory usage (you can use the provided system webview instead of bundling a Chromium in your game).
It makes perfect sense for an app that would need to be lightweight and easy on resources (chat app/productivity app).
For games, there is a bit less incentive on this side: you're usually playing just the game so you can afford putting more resources of your computer on it.

There is also the risk of getting some issues with outdated webviews or missing components on your players systems. Typically, Tauri uses WebKit via Edge on Windows (great) *or* MSHTML (aka IE10/11), which is a big no for anything :D 
In other words, we would need to ensure games are only run on system that have a good webview (I'm talking about windows here, but we should also check macOS and Linux).
This is usually less of a problem for browsers because they auto update so we can assume most players are on an up-to-date version.

**Reply by arthuro555** [reply]
> with Tauri what do we lose?

With Tauri, we lose node and its standard library. We have to use Tauri's APIs, and if we need something unsupported by it (as of now, it only supports files, native dialogs, native notifications, executing other processes and opening links in the default browser), we need to code the native part ourselves in rust. For GDevelop games that almost only use the filesystem on electron, this is not such a big issue, but if we wanted to port the GDevelop IDE like blurymind is suggesting, we would need to write our code in rust for for example Discord Rich Presence, the debugger server, streaming files from the filesystem to GDevelops build service, the external editors...

**Reply by arthuro555** [reply]
> Typically, Tauri uses WebKit via Edge on Windows (great) or MSHTML (aka IE10/11)

Also, I forgot to say that webkit is intelligent about using a webview, it [blocks MSHTML by default](https://github.com/tauri-apps/tauri-docs/issues/25#issuecomment-589941680). It actually uses Webview2, which has a fixed version mode of installation and an autoupdating shared evergreen version, allowing to have the recent features and/or stability electron provides.

**Comment by blurymind** [comment]
My understanding is that the main thing we get from electon is the file access without security restrictions. I wonder if some day we can completely replace electron with GD becoming a PWA

The new filesystem api might help us get there?
https://github.com/GoogleChromeLabs/text-editor

tauri looks interesting.

Honestly electron is a memory hog and eats up so much hd space. I would be happy if we find a better way to do this.

The external editors might need some work- they rely on it a bit too much now

**Reply by 4ian** [reply]
I think @arthuro555 was more talking about exported games than the GDevelop app itself.
It's also interesting to think about using it for GDevelop. It's right that Electron provides seamless access to the filesystem, but also a few other stuff, most notably 1) autoupdates and 2) the peace of mind of knowing that if it works on your machine, it works anywhere else because you have exactly the same runtime.

This being said, if we prove that we could re-implement missing features and that the webviews on all systems are of a quality high enough to support GDevelop like a ever-green recent browser would do, we could think about it :) 
An intermediate step would be to improve the PWA aspect of GDevelop. If it works well, it should be even easier to switch to an alternative runtime like Tauri.

**Reply by arthuro555** [reply]
Autoupdates is on the Tauri roadmap 👀

**Comment by arthuro555** [comment]
Tauri just went from alpha to beta and now has an autoupdater like electron 👀

**Comment by blurymind** [comment]
What about its debugging tools?

**Reply by 4ian** [reply]
The webview should provide its own debugging tools (https://tauri.studio/en/docs/usage/development/debugging) - which will depend on the system and what webview it is (not a lot of info in the doc it seems for now).
